builder/builder.go

```diff
@@ -0,0 +1,334 @@
+		cmder:        c,
+		debug:        debug,
+		workspaceDir: workspaceDir,
+		dryRun:       dryRun,
+		buildOptions: buildOptions,
+		mu:           sync.Mutex{},
+	}
+}
+
+// RunAllBuildSteps executes a pipeline.
+func (b *Builder) RunAllBuildSteps(ctx context.Context, dag *graph.Dag, pushTo []string) error {
+
+	if !b.dryRun {
+		if err := b.setupConfig(ctx); err != nil {
+			return err
+		}
+
+		if b.buildOptions.RegistryName != "" &&
+			b.buildOptions.RegistryPassword != "" &&
+			b.buildOptions.RegistryUsername != "" {
+
+			fmt.Printf("Logging in to registry: %s\n", b.buildOptions.RegistryName)
+			if err := b.dockerLoginWithRetries(ctx, 0); err != nil {
+				return err
+			}
+
+			fmt.Println("Successfully logged in")
+		}
+	}
+
+	root := dag.Nodes[graph.RootNodeID]
+	var completedChans []chan bool
+	errorChan := make(chan error)
+	for k, v := range dag.Nodes {
+		if k == graph.RootNodeID {
+			continue
+		}
+		completedChans = append(completedChans, v.Value.CompletedChan)
+	}
+
+	for _, n := range root.Children() {
+		go b.processVertex(ctx, dag, root, n, errorChan)
+	}
+
+	// Block until either:
+	// - The global context expires
+	// - A step has an error
+	// - All steps have been processed
+	for _, ch := range completedChans {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-ch:
+			continue
+		case err := <-errorChan:
+			return err
+		}
+	}
+
+	if b.buildOptions.Push {
+		if err := b.pushWithRetries(ctx, pushTo); err != nil {
+			return err
+		}
+	}
+
+	for k, v := range dag.Nodes {
+		if k == graph.RootNodeID {
+			continue
+		}
+
+		step := v.Value
+
+		err := b.PopulateDigests(ctx, step.ImageDependencies)
+		if err != nil {
+			return errors.Wrap(err, "failed to populate digests")
+		}
+
+		fmt.Printf("Step ID %v marked as %v (elapsed time in seconds: %f)\n", step.ID, step.StepStatus, step.EndTime.Sub(step.StartTime).Seconds())
+		bytes, err := json.Marshal(step.ImageDependencies)
+		if err != nil {
+			return errors.Wrap(err, "failed to unmarshal image dependencies")
+		}
+
+		fmt.Println(string(bytes))
+	}
+
+	return nil
+}
+
+// CleanAllBuildSteps cleans up all build steps and removes their corresponding Docker containers.
+func (b *Builder) CleanAllBuildSteps(ctx context.Context) {
+	// TODO: implement
+
+	// args := []string{"docker", "rm", "-f"}
+
+	// errs := cmder.Stop()
+	// 	if errs != nil && debug {
+	// 		fmt.Printf("Err during cleanup: %v", errs.String())
+	// 	}
+}
+
+func (b *Builder) processVertex(ctx context.Context, dag *graph.Dag, parent *graph.Node, child *graph.Node, errorChan chan error) {
+	err := dag.RemoveEdge(parent.Name, child.Name)
+	if err != nil {
+		errorChan <- errors.Wrap(err, "failed to remove edge")
+		return
+	}
+
+	degree := child.GetDegree()
+	if degree == 0 {
+		step := child.Value
+		step.StepStatus = graph.InProgress
+		step.StartTime = time.Now()
+		defer func() {
+			step.EndTime = time.Now()
+		}()
+
+		var args []string
+
+		if strings.HasPrefix(step.Run, "build") {
```

Comment:
>build [](start = 34, length = 5)

prefix "build " in case users use some commands like buildxyz

Reply:
fixed